src/peripherals.rs

```diff
@@ -202,30 +249,167 @@ where
         self.radio
             .shorts
             .modify(|_, w| w.disabled_rxen().disabled());
+
+        // We don't release the packet here because we may need to retransmit
     }
 
-    // TODO: Change to the bbqueue's Grants
     // Returns true if the ack was received successfully
+    // The upper stack is responsible for checking and disabling the timeouts
     pub(crate) fn check_ack(&mut self) -> bool {
         let ret = self.radio.crcstatus.read().crcstatus().is_crcok();
         // "Subsequent reads and writes cannot be moved ahead of preceding reads."
         compiler_fence(Ordering::Acquire);
-        // TODO: commit the grant if everything is ok
+
+        if ret {
+            if let (Some(tx_grant), Some(mut rx_grant)) =
+                (self.tx_grant.take(), self.rx_grant.take())
+            {
+                let pipe = tx_grant.pipe();
+                tx_grant.release();
+
+                let rssi = self.radio.rssisample.read().rssisample().bits();
+                rx_grant.set_pipe(pipe);
+                rx_grant.set_rssi(rssi);
+                rx_grant.commit_all();
+            } else {
+                unreachable!()
+            }
```

Comment:
@jamesmunns I used this pattern to avoid unwrapping, but I think `unreachable!()` could cause the same code bloat, what do you think ?

Reply:
unwrapping and panic are basically the same amount of overhead. If you are *reaaaaally* sure that it can never happen, you can use `unsafe { unreachable_unchecked() }`, but I'll take a bigger look later.

Reply:
It can happen if we mess something in the state machine code (`EsbIrq`), so I guess I will change it to `unwrap`.

Reply:
Alternatively, you could do something like this:

```rust
debug_assert!(tx_grant.is_some() && rx_grant.is_some());

let mut pipe = 255; // set to a known-bad value

if let Some(tx_grant) = self.tx_grant.take() {
    let pipe = tx_grant.pipe();
    tx_grant.release();
}

if let Some(mut rx_grant) = self.rx_grant.take() {
    let rssi = self.radio.rssisample.read().rssisample().bits();
    rx_grant.set_pipe(pipe);
    rx_grant.set_rssi(rssi);
    rx_grant.commit_all();
}
```

Which could still be seen in `release` mode, but is unlikely to cause any direct problems (unless someone uses it as an array index...)

Reply:
I decided to change it to `unwrap`, there are other places where we can't really recover, so we would need at least one panic, which would be enough to bring all the formatting machinery anyway. If one of those `unwraps` fails that means we have a bug somewhere.